### DISCUSSION TITLE
Voice is not quite understandable

### DISCUSSION THREAD

**Discussion by LtSnakePlissken** [judahpaul16/gpt-home#55] [Q&A]
Hey guys, I'm just wondering if anyone else is having issues understanding the voice? Is it possible to configure the voice somehow or use a different voice service? I'm not quite able to understand most of the time. 

This thing is amazing by the way!

**Comment by judahpaul16** [comment]
Hey I'm glad you found the project useful!

In `pyttsx3` on Linux, the default voices are provided by eSpeak, and they include a variety of languages and accents including female voices and even supports custom voice files.  

Within `common.py`:
```python
# ... engine = pyttsx3.init()

# Set the desired voice (example: setting American English voice)
engine.setProperty('voice', 'en-us')

# ... (rest of the file)
```

after modifying the file you may have to restart the app with `gpt-restart` or `docker exec -it gpt-home supervisorctl restart app`.

For more details, see the eSpeak documentation [here](https://espeak.sourceforge.net/voices.html) and [here](https://espeak.sourceforge.net/languages.html).

If you want to train a totally custom voice like your own voice, you can check out [TextyMcSpeechy](https://github.com/domesticatedviking/TextyMcSpeechy) by @domesticatedviking.

**Reply by judahpaul16** [reply]
Alternatively, you can opt to use a different TTS engine by replacing the `pyttsx3` setup in `common.py` with your preferred TTS service. 
```python
# Initialize the text-to-speech engine
engine = pyttsx3.init()
# Set properties
engine.setProperty('rate', 145)
engine.setProperty('volume', 1.0)
# Direct audio to specific hardware
engine.setProperty('alsa_device', 'hw:Headphones,0')
```

You would invoke that service's version of `engine.say()` in the `speak()` function:

```python
async def speak(text, stop_event=asyncio.Event()):
    async with speak_lock:
        loop = asyncio.get_running_loop()
        def _speak():
            engine.say(text)
            engine.runAndWait()
        await loop.run_in_executor(executor, _speak)
        stop_event.set()
```

Example using [`piper`](https://github.com/rhasspy/piper):

```python
# Initialize Piper voice model and device
voice_model = 'your_voice_model'  # Replace with your specific voice model identifier
device = 'hw:Headphones,0'

async def speak(text, stop_event=asyncio.Event()):
    async with speak_lock:
        loop = asyncio.get_running_loop()
        def _speak():
            command = ['piper', '--voice', voice_model, '-o', device, '--text', text]
            subprocess.run(command)
        await loop.run_in_executor(executor, _speak)
        stop_event.set()
```

**Comment by judahpaul16** [comment]
GPT Home [`v1.7.0`](https://github.com/judahpaul16/gpt-home/releases/tag/v1.7.0) now supports [`gTTS`](https://github.com/pndurette/gTTS) for a more realistic sounding voice. Just change it in the settings tab at [`gpt-home.local/settings`](gpt-home.local/settings) or the IP address of the raspberry pi.

![image](https://github.com/user-attachments/assets/35fc0ca7-40b5-40bd-8d1b-c8a2d83e25a3)
